runtime/lua/vim/snippet.lua

```diff
@@ -0,0 +1,286 @@
+--- - When a variable isn't set, return its default (if any) or an empty string.
+---
+--- Note that in some cases, the default is ignored since it's not clear how to distinguish an empty
+--- value from an unset value (e.g.: `TM_CURRENT_LINE`).
+---
+--- @param var string
+--- @param default string
+--- @return string?
+local function resolve_variable(var, default)
+  --- @param str string
+  --- @return string
+  local function expand_or_default(str)
+    local expansion = vim.fn.expand(str) --[[@as string]]
+    return expansion == '' and default or expansion
+  end
+
+  if var == 'TM_SELECTED_TEXT' then
+    return default
+  elseif var == 'TM_CURRENT_LINE' then
+    return vim.api.nvim_get_current_line()
+  elseif var == 'TM_CURRENT_WORD' then
+    return expand_or_default('<cword>')
+  elseif var == 'TM_LINE_INDEX' then
+    return tostring(vim.fn.line('.') - 1)
+  elseif var == 'TM_LINE_NUMBER' then
+    return tostring(vim.fn.line('.'))
+  end
+
+  if var == 'TM_FILENAME' then
+    return expand_or_default('%:t')
+  elseif var == 'TM_FILENAME_BASE' then
+    -- Not using '%:t:r' since we want to remove all extensions.
+    local filename_base = expand_or_default('%:t'):gsub('%.[^%.]*$', '')
+    return filename_base
+  elseif var == 'TM_DIRECTORY' then
+    return expand_or_default('%:p:h:t')
+  elseif var == 'TM_FILEPATH' then
+    return expand_or_default('%:p')
+  end
+
+  -- Unknown variable.
+  return nil
+end
+
+--- Removes newlines from the given text.
+---
+--- @param text string|string[]
+--- @return string[]
+local function remove_newlines(text)
+  text = type(text) == 'string' and { text } or text
+  --- @cast text string[]
+  return vim.split(table.concat(text), '\n')
+end
+
+--- Appends the given text to the snippet, taking care of indentation.
+---
+--- @param snippet string[]
+--- @param text string|string[]
+local function append_to_snippet(snippet, text)
+  -- Get the base indentation based on the current line and the last line of the snippet.
+  local base_indent = vim.api.nvim_get_current_line():match('^%s*') or ''
+  if #snippet > 0 then
+    base_indent = base_indent .. (snippet[#snippet]:match('^%s*') or '') --- @type string
+  end
+
+  local expand_tab = vim.api.nvim_get_option_value('expandtab', {})
+  local lines = vim.iter.map(function(i, line)
+    -- Replace tabs by spaces.
+    if expand_tab then
+      line = line:gsub('\t', (' '):rep(vim.fn.shiftwidth())) --- @type string
+    end
+    -- Add the base indentation.
+    if i > 1 then
+      line = base_indent .. line
+    end
+    return line
+  end, ipairs(remove_newlines(text)))
+
+  table.insert(snippet, table.concat(lines, '\n'))
+end
+
+--- Computes the 0-based position of a tabstop located at the end of `snippet` and spanning
+--- `placeholder` (if given).
+--- @param cursor_row integer
+--- @param snippet string[]
+--- @param placeholder string?
+--- @return Range4
```

Comment:
I didn't want to define yet another range type and so I opted to use `Range4` here. It might be confusing given that the definition lives in a treesitter module, but it also seems like there's a spike in refactoring the entire position/range API (#25509) so I'm unsure of what would be the best action right now.

Reply:
Right, but here I'm not dealing with ranges as defined by the LSP spec (I'm not receiving or sending LSP objects from the client directly), this is just for me to be able to represent the range of a snippet tabstop in a less verbose way.